[PR-16095] Add documentation for shopify coupon auto-sync (#249)
talkable/talkable-docs: source/advanced_features/shopify_coupons_auto_sync.rst

.. _advanced_features/shopify_coupons_auto_sync:

.. meta::
  :description: Talkable supports coupon auto-sync for Shopify. This feature allows to avoid manual coupon uploads and/or Shopify Price Rule management.

> Shopify coupon auto-sync
> =========================

Reviewer: ```suggestion
Shopify coupon auto-sync
========================
```

Talkable supports coupon auto-sync for Shopify. This feature allows to avoid manual coupon uploads and/or Shopify Price Rule management.

Get started
-----------

Before using coupon auto-sync, you have to install our Shopify app (Settings → Shopify Integration → click Authorize).

> To enable coupon auto-sync in a coupon list, go to coupon list edit/create page and check “Enable Coupon List Shopify Auto Sync”. If this checkbox is disabled, it means this site uses Coupon Webhook, and no coupons can be directly uploaded to Shopify. Instead, coupons are delivered to Coupon Webhook url, and that’s where they should be handled.

Reviewer: ```suggestion
To enable coupon auto-sync in a coupon list, go to coupon list edit/create page and check “Enable Coupon List Shopify Auto Sync”. If this checkbox is disabled, it means this site uses Coupon Webhook, and no coupons can be directly uploaded to Shopify. Instead, coupons are delivered to Coupon Webhook URL, and that’s where they should be handled.
```

.. note::

>   Talkable uses Shopify Price Rule ID (if present) to determine where to upload newly generated coupons. It is recommended to leave it blank when creating coupon lists. In that case Talkable will create a Price Rule based on coupon list configuration and store its ID in coupon list. Please see **Advanced features** for more information about Shopify Price Rule ID.

Reviewer: ```suggestion
  Talkable uses Shopify Price Rule ID (if present) to determine where to upload newly generated coupons. It is recommended to leave it blank when creating coupon lists. In that case, Talkable will create a Price Rule based on coupon list configuration and store its ID in the coupon list. Please see **Advanced features** for more information about Shopify Price Rule ID.
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Talkable applies coupon list settings onto Price Rule:

.. container:: ptable

  ===================== =================================== ======================================
  Coupon list attribute Price rule attribute (as in API)    Price rule attribute (as in interface)
  ===================== =================================== ======================================
  **Amount**            `value`                             Value
  **Discount type**     `value_type`                        Type
  **Expiration date**   `ends_at`                           End date
  **Minimum subtotal**  `prerequisite_subtotal_range`       Minimum purchase amount
  ===================== =================================== ======================================

Some other Price Rule attributes that we set when generating Price Rule:

.. container:: ptable

  ======================== ====================================== ========================
  Price Rule attribute     Value                                  Attribute effect
  ======================== ====================================== ========================

>   `title`                  e.g. "Talkable single-use coupons $10" Describes Price Rule.

Reviewer 2: Is there something we can do to make the first and the second columns fit values in them? Currently, items like `customer_selection` are split across lines.

  `usage_limit`            `1`                                    Each discount code can be used exactly once.
  `customer_selection`     `all`                                  The price rule is valid for all customers.
  `starts_at`              Time of coupon list creation           The date and time when the price rule starts.
  ======================== ====================================== ========================

Why aren’t there any coupons created for my auto-synced list?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Auto-sync generates and uploads coupons when there is demand for them. That means that nothing will happen before there are rewards waiting for coupons from this coupon list. When there is at least one such reward, Talkable generates a code and syncs it with Shopify, while scheduling a refill task that will generate a bunch of codes in advance. This logic is triggered every time a coupon is issued, so there always should be some available coupons for future rewards.

Advanced features
-----------------

Shopify Price Rule ID
~~~~~~~~~~~~~~~~~~~~~

> Talkable allows to assign custom Price Rule ID to a coupon list. This might be helpful if clients want to deviate from our default Price Rule settings, e.g. restrict price rule to a group of customer or kind of products.

Reviewer: ```suggestion
Talkable allows to assign custom Price Rule ID to a coupon list. This might be helpful if clients want to deviate from our default Price Rule settings, e.g. restrict price rule to a group of customers or kinds of products.
```

It is recommended to create a new Price Rule in Shopify if there is a need to use a custom one, instead of editing the Price Rule generated by Talkable.

.. warning::

>   Talkable validates Price Rule attached to coupon list. There are certain Price Rule attributes that must match coupon list configuration.

Reviewer: ```suggestion
  Talkable validates Price Rule attached to the coupon list. There are certain Price Rule attributes that must match coupon list configuration.
```

  These include:

  `value` - must correspond to coupon list amount

  `value_type` - must correspond to coupon list type

  `ends_at` - must be greater than or equal to coupon list expiration (and absent if coupon list has no expiration)

>   `prerequisite_subtotal_range` - must match coupon list minimum subtotal

Reviewer 2: I'm not sure in what order yours and Taras' code is going to be released, but in Taras' feature `prerequisite_subtotal_range` will no longer be validated.

Author: @Doimhneacht do you mean this task PR https://github.com/talkable/talkable/pull/7601 ?

Reviewer 2: @max-si-m Precisely.

If Price Rule passes validation, it can be attached to a coupon list and will be used as a new destination for all the coupons created from that moment on.

.. note::
  All the previously generated coupons will remain in the Price Rule that was used when they were auto-synced.

Shopify Price Rule Changed Email Notification
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You might modify Price Rules on Shopify by mistake, making them incompatible with the coupon lists they are attached to. In order to find out about such changes as early as possible, we have a daily monitoring job that checks that Price Rules have no critical differences from respective coupon lists.

Attributes that are checked in this job are the following:

  - `usage_limit` - must always be 1

  - `value` - must correspond to coupon list amount

  - `value_type` - must correspond to coupon list type

  - `ends_at` - must be greater than or equal to coupon list expiration (and absent if coupon list has no expiration)

>   - `prerequisite_subtotal_range` - must match coupon list minimum subtotal

Reviewer 2: Please consult with @dgtv if this is still true after his update of the letter. I believe `prerequisite_subtotal_range` is never included in the letter, but I am not entirely sure how the `ends_at` is supposed to work. Sorry, discussed it a long time ago.

If any of these attributes differ from what they are expected to be and Talkable cannot fix that by updating a coupon list (see **Coupon list sync**), Talkable sends an email notification.

> Once Price Rule becomes critically different from the coupon list it is assigned to, the coupon list is no longer editable. Please encourage clients to fix issues listed in the email notification to remedy this situation.

Reviewer: ```suggestion
Once Price Rule becomes critically different from the coupon list it is assigned to, the coupon list is no longer editable. Please encourage clients to fix the issues listed in the email notification to remedy this situation.
```

Coupon list sync
~~~~~~~~~~~~~~~~

Talkable tries to keep up with the Price Rules assigned to coupon lists when/if Price Rules change.

As long as the Price Rule is otherwise valid for a coupon list, we update the coupon list’s:

**expiration date** - only if Price Rule end date is further in the future (or absent)

**minimum subtotal**

.. note::

>   If there are any other changes in Price Rule that make it not suitable for a certain coupon list, we won’t sync coupon list. In this case a Shopify Price Rule Changed Email Notification will be delivered and actions from clients will be required to fix the issue.

Reviewer: ```suggestion
  If there are any other changes in Price Rule that make it not suitable for a certain coupon list, we won’t sync coupon list. In this case, a Shopify Price Rule Changed Email Notification will be delivered and actions from clients will be required to fix the issue.
```

> This sync is performed on a daily basis. Do not expect immediate effect after Price Rule update.

Reviewer: Shouldn’t this be part of a `note::`?

```suggestion
This sync is performed daily. Do not expect immediate effect after Price Rule update.
```
